### PR TITLE
Optimistic UI for datetimes | Closes #2319, #2320

### LINKED ISSUE #2319: Prepare queries for optimistic responses
As mentioned in [this comment](https://github.com/eventespresso/event-espresso-core/issues/2314#issuecomment-581276289), we need to update queries that use `client.readQuery` to make use of `useQuery` to be able to get updated when cache gets updated by optimistic responses.

### PR DESCRIPTION
**Pull request by manzoorwanijk**
This PR:
- Removes the usage of `client.readQuery` from query hooks
- Creates `useCacheQuery` hook for reading cache to properly set subscriptions for optimistic responses.
- Creates optimistic responses for datetime mutations
- Updates all the tests affected by using `useCacheQuery` and optimistic responses.
- Adds improvements to card components.
- Adds `/assets/dist` to `.gtignore`
- Closes #2319, #2320

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -1,3 +0,0 @@
-{
```
Comment (nicolad) [inline .vscode/settings.json]: :crying_cat_face:

File: .vscode/extensions.json [hunk 2]
```diff
@@ -1,3 +0,0 @@
-{
```
Comment (nicolad) [inline .vscode/extensions.json]: :cry:

File: assets/prototype/domain/eventEditor/datetimes/dateCard/DateCard.tsx [hunk 3]
```diff
@@ -27,18 +27,21 @@ const DateCard: React.FC<ListItemProps> = ({ id }): JSX.Element => {
 	const { updateEntity } = useEntityMutator(EntityType.Datetime, id);
 	const { getRelations } = useRelations();
 
+	const startDate = parseISO(date.startDate) || PLUS_ONE_MONTH;
+	const endDate = parseISO(date.endDate) || PLUS_TWO_MONTHS;
```
Comment (tn3rb) [inline assets/prototype/domain/eventEditor/datetimes/dateCard/DateCard.tsx]: 👍

File: assets/prototype/domain/eventEditor/tickets/ticketCard/TicketCard.tsx [hunk 4]
```diff
@@ -39,14 +39,17 @@ const TicketCard: React.FC<ListItemProps> = ({ id }): JSX.Element => {
...
-							defaultValue={ticket.name}
 							value={ticket.name}
 							onCancel={(value: any): void => {
 								console.log('TicketProvider title onCancel => NEEDS CALLBACK');
 								console.log('value', value);
 							}}
-							onConfirm={(name: string): MutationResult => updateEntity({ name })}
-							minWidth={'320px'}
+							onConfirm={(name: string): void => {
+								if (name !== ticket.name) {
```
Comment (tn3rb) [inline assets/prototype/domain/eventEditor/tickets/ticketCard/TicketCard.tsx]: 👍

File: assets/prototype/domain/eventEditor/data/entities/test/useUpdateDatetimeList.test.ts [hunk 5]
```diff
@@ -8,10 +8,11 @@ import useDatetimes from '../../queries/datetimes/useDatetimes';
 import useDatetimeIds from '../../queries/datetimes/useDatetimeIds';
 import { ApolloMockedProvider } from '../../../context/TestContext';
 
+const timeout = 5000; // milliseconds
```
[the remark this answers is not in the document]
Comment (nicolad) [reply assets/prototype/domain/eventEditor/data/entities/test/useUpdateDatetimeList.test.ts]: We can have something as sophisticated as `beforeAll`: https://jestjs.io/docs/en/setup-teardown#one-time-setup
or a simple shared test-only constant

[the remark this answers is not in the document]
Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/data/entities/test/useUpdateDatetimeList.test.ts]: Actually it's not the jest timeout, it's the timeout that we need for `renderHook` wait functions. We can may be define it somewhere and import everywhere.

File: assets/prototype/domain/eventEditor/data/mutations/datetimes/useDatetimeMutator.ts [hunk 6]
```diff
@@ -44,12 +70,63 @@ const useDatetimeMutator = (): Mutator => {
 		};
 	};
 
+	const createOptimisticResponse = (mutationType: MutationType, input: MutationInput): any => {
```
[the remark this answers is not in the document]
Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/data/mutations/datetimes/useDatetimeMutator.ts]: Agreed

[the remark this answers is not in the document]
Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/data/mutations/datetimes/useDatetimeMutator.ts]: Done

**Pull request merged**